internal/ingress/status/status.go

```diff
@@ -89,109 +79,35 @@ type Config struct {
-
-	var cancelContext context.CancelFunc
-
-	var newLeaderCtx = func(ctx context.Context) context.CancelFunc {
-		// allow to cancel the context in case we stop being the leader
-		leaderCtx, cancel := context.WithCancel(ctx)
-		go s.elector.Run(leaderCtx)
-		return cancel
-	}
-
-	var stopCh chan struct{}
-	callbacks := leaderelection.LeaderCallbacks{
-		OnStartedLeading: func(ctx context.Context) {
-			klog.V(2).Infof("I am the new status update leader")
-			stopCh = make(chan struct{})
-			go s.syncQueue.Run(time.Second, stopCh)
-			// trigger initial sync
-			s.syncQueue.EnqueueTask(task.GetDummyObject("sync status"))
-			// when this instance is the leader we need to enqueue
-			// an item to trigger the update of the Ingress status.
-			wait.PollUntil(updateInterval, func() (bool, error) {
-				s.syncQueue.EnqueueTask(task.GetDummyObject("sync status"))
-				return false, nil
-			}, stopCh)
-		},
-		OnStoppedLeading: func() {
-			klog.V(2).Info("I am not status update leader anymore")
-			close(stopCh)
-
-			// cancel the context
-			cancelContext()
-
-			cancelContext = newLeaderCtx(ctx)
-		},
-		OnNewLeader: func(identity string) {
-			klog.Infof("new leader elected: %v", identity)
-		},
-	}
-
-	broadcaster := record.NewBroadcaster()
-	hostname, _ := os.Hostname()
-
-	recorder := broadcaster.NewRecorder(scheme.Scheme, apiv1.EventSource{
-		Component: "ingress-leader-elector",
-		Host:      hostname,
-	})
+// Start starts the loop to keep the status in sync
+func (s statusSync) Run(stopCh chan struct{}) {
+	go s.syncQueue.Run(time.Second, stopCh)
 
-	lock := resourcelock.ConfigMapLock{
-		ConfigMapMeta: metav1.ObjectMeta{Namespace: s.pod.Namespace, Name: electionID},
-		Client:        s.Config.Client.CoreV1(),
-		LockConfig: resourcelock.ResourceLockConfig{
-			Identity:      s.pod.Name,
-			EventRecorder: recorder,
-		},
-	}
+	// trigger initial sync
+	s.syncQueue.EnqueueTask(task.GetDummyObject("sync status"))
 
-	ttl := 30 * time.Second
-	var err error
-	s.elector, err = leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
-		Lock:          &lock,
-		LeaseDuration: ttl,
-		RenewDeadline: ttl / 2,
-		RetryPeriod:   ttl / 4,
-		Callbacks:     callbacks,
-	})
-	if err != nil {
-		klog.Fatalf("unexpected error starting leader election: %v", err)
-	}
-
-	cancelContext = newLeaderCtx(ctx)
+	// when this instance is the leader we need to enqueue
+	// an item to trigger the update of the Ingress status.
+	wait.PollUntil(updateInterval, func() (bool, error) {
+		s.syncQueue.EnqueueTask(task.GetDummyObject("sync status"))
+		return false, nil
+	}, stopCh)
 }
 
-// Shutdown stop the sync. In case the instance is the leader it will remove the current IP
+// Shutdown stops the sync. In case the instance is the leader it will remove the current IP
 // if there is no other instances running.
 func (s statusSync) Shutdown() {
 	go s.syncQueue.Shutdown()
 
-	// remove IP from Ingress
-	if s.elector != nil && !s.elector.IsLeader() {
```

Comment:
@aledbf - do you know why this code was removed. we are running into a regression where during rolling update the ingress status is getting removed by the last pod of old deployment and we are suspecting this is the issue here.

Reply:
@rajatjindal because we now use a callback https://github.com/kubernetes/ingress-nginx/pull/3885/files#diff-7d2578dc5d43f0b395d7cd3312c5c489R63

> the ingress status is getting removed by the last pod of old deployment and we are suspecting this is the issue here.

This is expected. If you don't want this behavior you need to set UpdateStatusOnShutdown=false (flag `--update-status-on-shutdown`)